PEP 004 - Labels and Namespaces

carolynvs opened the pull request:
Organize and isolate porter documents using labels and namespaces.

This lays storage foundation that PEP003 relies upon.

### New PEP Checklist
- [x] Is your pull request named: `PEP NNN - <TITLE>`?
- [x] Have you posted your proposal in the [forum](https://porter.sh/forum)?
- [x] Have you gotten approval and a PEP number from a maintainer?
- [x] Did you use the [PEP Template](https://github.com/getporter/proposals/blob/master/pep/000-PEP-TEMPLATE.md)?

#### Forum Suggestion Link


Originally proposed as part of PEP 003: Dependency Namespaces and Labels. https://github.com/getporter/proposals/pull/8

vdice commented on pep/004-labels-and-namespaces.md:
````diff
@@ -0,0 +1,386 @@
+}
+```
+
+During upgrade, invoke and uninstall the user-defined labels are applied to the claim only:
````
Curious around the design decision here around _not_ updating the installation with the user-defined label(s) for these actions?  (And we might consider adding that decision to the doc for context...)

carolynvs replied:
I think I'm trying to put too much into the install command here because I wanted to avoid either of the situations below

**applying installation labels separate from the installation creation**
```
porter install wordpress
porter installation label set owner=carolynvs
```

**adding weird one-off flags to install**
```
porter install wordpress --installation-label owner=carolynvs
```

But what I suggested in the proposal is really inconsistent (sometimes --label applies to an installation label and sometimes a claim label)...

I guess out of all these options, I would go with "you only set installation labels using the porter installation label command" and not try to pack a unique behavior into install.

vdice replied:
Yea, the inconsistency might feel weird.  I'm still unclear on why we wouldn't want to, say, `porter upgrade wordpress --label owner=carolynvs` to update the installation as well?  If I would be able to run `porter upgrade wordpress && porter installation label set wordpress owner=carolynvs`, why not allow for the inline `--label` on any action to take effect (modify the installation as well as the claim)?

carolynvs replied:
I guess I'm getting hung up on being able to set labels on the claim and the installation independently, you know?

In my head I was imagining that it could be useful to put a label _just_ on the claim, to communicate something about that particular run of the bundle. For example, the operator could add the label "porter.sh/operator-revision=REVISION_FROM_KUBERNETES_THAT_TRIGGERED_THE_RUN". I really wouldn't want that stored on the installation as well, it should just be on the claim.

carolynvs replied:
I'm going to redo this with `porter install --label-installation KEY=VALUE`. I think that's the most clear out of all the options. Then `--label` on all of the actions always maps to the claim labels. After the installation exists, people can use `porter installation label set` to manage the labels.

vdice commented on pep/004-labels-and-namespaces.md:
```diff
@@ -0,0 +1,386 @@
+
+#### Output Labels
+
+The spec allows for labels to be set on a result however we will not expose that in Porter.
```
These two lines appear to be duplicates of the lines under `Result Labels` above.  Intended to repeat?

carolynvs replied:
copy/paste strikes again. I'll make it unique and cleanup.

carolynvs commented:
@vdice I have addressed your feedback and synced this with the open PR to the cnab-spec. If you think this is good enough to start iterating on the PEP design, let's merge it. The goal with the PEPs is that merging the first PR as "provisional" means that we want to do the work but haven't finalized the design. I'm going to start relying on that more so that we can get initial proposals out for people to see without requiring it to be complete before merging the provisional PEP.

Once we finish design and are ready to implement, I'll update the PEP to say "implementable".

vdice reviewed (approved):
LGTM
